lighthouse-core/fraggle-rock/user-flow.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const {navigation, startTimespan, snapshot} = require('./api.js');
+
+/** @typedef {{config?: LH.Config.Json, configContext?: LH.Config.FRContext}} UserFlowOptions */
+/** @typedef {UserFlowOptions & {page: import('puppeteer').Page}} FrOptions */
+/** @typedef {UserFlowOptions & {stepName?: string}} StepOptions */
+
+class UserFlow {
+  /**
+   * @param {import('puppeteer').Page} page
+   * @param {UserFlowOptions=} options
+   */
+  constructor(page, options) {
+    /** @type {FrOptions} */
+    this.options = {page, ...options};
+    /** @type {LH.FlowResult.Step[]} */
+    this.steps = [];
+  }
+
+  /**
+   * @param {string} url
+   * @param {StepOptions=} overrideOptions
+   */
+  async navigate(url, overrideOptions) {
+    const options = {url, ...this.options, ...overrideOptions};
+    const result = await navigation(options);
+    if (!result) throw Error('Navigation returned undefined');
+    this.steps.push({
+      lhr: result.lhr,
+      name: overrideOptions && overrideOptions.stepName,
+    });
+  }
+
+  /**
+   * @param {StepOptions=} overrideOptions
+   */
+  async startTimespan(overrideOptions) {
+    const options = {...this.options, ...overrideOptions};
+    const timespan = await startTimespan(options);
+    if (this.currentTimespan) throw Error('Timespan already in progress');
+    this.currentTimespan = {timespan, options};
+  }
+
+  async endTimespan() {
+    if (!this.currentTimespan) throw Error('No timespan in progress');
+    const {timespan, options} = this.currentTimespan;
+    const result = await timespan.endTimespan();
+    this.currentTimespan = undefined;
+    if (!result) throw Error('Timespan returned undefined');
+    this.steps.push({
+      lhr: result.lhr,
+      name: options && options.stepName,
+    });
+  }
+
+  /**
+   * @param {StepOptions=} overrideOptions
+   */
+  async snapshot(overrideOptions) {
+    const options = {...this.options, ...overrideOptions};
+    const result = await snapshot(options);
+    if (!result) throw Error('Snapshot returned undefined');
+    this.steps.push({
+      lhr: result.lhr,
+      name: overrideOptions && overrideOptions.stepName,
+    });
+  }
+
+  /**
+   * @return {LH.FlowResult}
+   */
+  getFlowResult() {
```

Comment:
WDYT about exposing the artifacts/report too? Similar to our node result today where you get `lhr`, `report`, `artifacts`

`return {flhr, report, artifacts}` and/or include artifacts as optional on steps?

Reply:
SGTM